[ALLUXIO-2434] UFS Abstraction

madanadit opened the pull request:
Summary of changes:
- UnderFileSystem -> interface instead of abstract class
- HDFS and local UFS derive from BaseUnderFileSystem 
- Object stores derive from ObjectUnderFileSystem

Plugging in new object store requires significantly less boiler plate code. Abstract ObjectUnderFileSystem handles directory representation. Implementations need to implement the following new methods instead of existing:
- createObject
- copyObject
- deleteObject
- getObjectStatus
- getFolderSuffix
- getObjectListing
- putObject

JIRA: https://alluxio.atlassian.net/browse/ALLUXIO-2434

madanadit commented:
@calvinjia @gpang Please take a look. Thanks!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12400/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12400/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12400/org.alluxio$alluxio-tests/testReport/alluxio.master.file/FileSystemMasterIntegrationTest/concurrentDelete/'><strong>alluxio.master.file.FileSystemMasterIntegrationTest.concurrentDelete</strong></a></li></ul>
Test FAILed.

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,639 @@
+   * @throws IOException if failed to create stream
+   * @return new OutputStream
+   */
+  protected abstract OutputStream createObject(String key) throws IOException;
```
Can this be collapsed into `createDirect` after directories are removed?

calvinjia replied:
Sure, let's revisit this when the full change is complete.

calvinjia commented on underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java:
```diff
@@ -458,59 +148,105 @@ private boolean copy(String src, String dst) {
+   * Wrapper over GCS {@link StorageObjectsChunk}.
    */
-  private ObjectMetadata getObjectDetails(String key) {
+  final class GCSObjectListingResult implements ObjectListingResult {
```
Should this be `OSSObjectListingResult`? Similar comments to `GCSObjectListingResult`

calvinjia commented on underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java:
```diff
@@ -458,59 +148,105 @@ private boolean copy(String src, String dst) {
+  @Override
+  protected ObjectListingResult getObjectListing(String path, boolean recursive)
+      throws IOException {
+    return new GCSObjectListingResult(path, recursive);
```
Should this be `OSSObjectListingResult`?

calvinjia commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java:
```diff
@@ -572,251 +286,121 @@ private boolean copy(String src, String dst) {
-    } catch (AmazonClientException e) {
-      // Its possible that the object is not a file but a folder
-      return getFolderMetadata(stripPrefixIfPresent(key));
+  final class S3ObjectListingResult implements ObjectListingResult {
```
Should we distinguish between the `S3ObjectListingResults`?

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -475,34 +296,7 @@ public OutputStream create(String path, CreateOptions options) throws IOExceptio
-    }
-    return returnPaths.toArray(new String[returnPaths.size()]);
-  }
+  String[] listRecursive(String path) throws IOException;
```
instead of introducing a method for recursive listing, introduce an options object

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -618,7 +405,7 @@ public void setProperties(Map<String, String> properties) {
    * @throws IOException if a non-Alluxio error occurs
    */
-  public abstract String getOwner(String path) throws IOException;
+  String getOwner(String path) throws IOException;
```
sort interface methods alphabetically

madanadit replied:
done

jsimsa commented:
@madanadit thanks for this awesome PR ... left some comments on the UnderFileSystem interface

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12403/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12403/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12403/org.alluxio$alluxio-tests/testReport/alluxio.master.file/FileSystemMasterIntegrationTest/concurrentDelete/'><strong>alluxio.master.file.FileSystemMasterIntegrationTest.concurrentDelete</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12404/<h2>Failed Tests: <span class='status-failure'>37</span></h2><h3><a name='org.alluxio:alluxio-core-client' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport'>org.alluxio:alluxio-core-client</a>: <span class='status-failure'>16</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/singleByteWrite/'><strong>alluxio.client.file.FileOutStreamTest.singleByteWrite</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/manyBytesWrite/'><strong>alluxio.client.file.FileOutStreamTest.manyBytesWrite</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/writeOffset/'><strong>alluxio.client.file.FileOutStreamTest.writeOffset</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/cancelWithDelegation/'><strong>alluxio.client.file.FileOutStreamTest.cancelWithDelegation</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/cancelWithoutDelegation/'><strong>alluxio.client.file.FileOutStreamTest.cancelWithoutDelegation</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/cacheWriteExceptionNonSyncPersist/'><strong>alluxio.client.file.FileOutStreamTest.cacheWriteExceptionNonSyncPersist</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/cacheWriteExceptionSyncPersist/'><strong>alluxio.client.file.FileOutStreamTest.cacheWriteExceptionSyncPersist</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/truncateWrite/'><strong>alluxio.client.file.FileOutStreamTest.truncateWrite</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/writeBadBufferOffset/'><strong>alluxio.client.file.FileOutStreamTest.writeBadBufferOffset</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/writeNullBuffer/'><strong>alluxio.client.file.FileOutStreamTest.writeNullBuffer</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/writeNullBufferOffset/'><strong>alluxio.client.file.FileOutStreamTest.writeNullBufferOffset</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/asyncWrite/'><strong>alluxio.client.file.FileOutStreamTest.asyncWrite</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/getBytesWrittenWithDifferentUnderStorageType/'><strong>alluxio.client.file.FileOutStreamTest.getBytesWrittenWithDifferentUnderStorageType</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/close/'><strong>alluxio.client.file.FileOutStreamTest.close</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/flush/'><strong>alluxio.client.file.FileOutStreamTest.flush</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-client/testReport/alluxio.client.file/FileOutStreamTest/writeBuffer/'><strong>alluxio.client.file.FileOutStreamTest.writeBuffer</strong></a></li></ul><h3><a name='org.alluxio:alluxio-core-server' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport'>org.alluxio:alluxio-core-server</a>: <span class='status-failure'>21</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getOutputStream/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getOutputStream</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/completeNonExistentUfsFile/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.completeNonExistentUfsFile</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/completeUfsFileInvalidSession/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.completeUfsFileInvalidSession</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/cancelNonExistentUfsFile/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.cancelNonExistentUfsFile</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/cancelUfsFileInvalidSession/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.cancelUfsFileInvalidSession</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/openUfsFile/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.openUfsFile</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/openNonExistentUfsFile/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.openNonExistentUfsFile</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/closeUfsFile/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.closeUfsFile</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/closeNonExistentUfsFile/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.closeNonExistentUfsFile</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/closeUfsFileInvalidSession/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.closeUfsFileInvalidSession</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getNonExistentOutputStream/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getNonExistentOutputStream</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getInputStreamAtPosition/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getInputStreamAtPosition</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getInputStreamAtPositionOptimized/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getInputStreamAtPositionOptimized</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getInputStreamAtPositionCache/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getInputStreamAtPositionCache</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getInputStreamAtPositionInvalidCache/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getInputStreamAtPositionInvalidCache</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getInputStreamAtEOF/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getInputStreamAtEOF</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getNonExistentInputStream/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getNonExistentInputStream</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/cleanSessions/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.cleanSessions</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/createUfsFile/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.createUfsFile</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/cancelUfsFile/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.cancelUfsFile</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12404/org.alluxio$alluxio-core-server/testReport/alluxio.worker.file/UnderFileSystemManagerTest/getInputStream/'><strong>alluxio.worker.file.UnderFileSystemManagerTest.getInputStream</strong></a></li></ul>
Test FAILed.

madanadit commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,639 @@
+    ObjectListingResult chunk = getObjectListing(path, recursive);
+    chunk = chunk.getNextChunk();
+    if (chunk == null) {
+      return null;
```
chunk is null if 
- if there list didn't find anything with prefix path (initially)
- or, a chunk was processed and there are no more chunks remaining

the 2 places where getObjectListing is called were passing in a path which had been normalized with PATH_SEPARATOR. I changed it so that the implementation normalizes it if not already so. the return should be null if path is a file. i need to add test cases for this class.

madanadit commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,639 @@
+
+  @Override
+  public OutputStream create(String path, CreateOptions options) throws IOException {
+    return createDirect(path, options);
```
Let me address that in a later PR.

madanadit commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,639 @@
+   *
+   * @param child the key of the child
+   * @param parent the key of the parent
+   * @return the child key with the parent prefix removed, null if the parent prefix is invalid
```
No. I changed to throwing an exception.

madanadit commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -377,7 +200,7 @@ public OutputStream create(String path, CreateOptions options) throws IOExceptio
    * @throws IOException if a non-Alluxio error occurs
    */
-  public abstract List<String> getFileLocations(String path, long offset) throws IOException;
+  List<String> getFileLocations(String path, long offset) throws IOException;
```
done

madanadit commented:
@calvinjia @jsimsa Thanks for the review!. I have addressed most of the comments. The ones left are:
- Calvin's comment on ObjectListingResult [will talk offline]
- Jiri's comment on consistently using the options idiom for createDirect and listRecursive [will create another PR]

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12405/
Test PASSed.

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileStatus.java:
```diff
@@ -66,6 +66,9 @@ public String toString() {
    */
   public static String[] toListingResult(UnderFileStatus[] children) {
+    if (children == null) {
+      return null;
```
would it make sense to return an empty array instead? that way the caller does not have to test for null

madanadit replied:
A null return is used to detect a non-existent directory; returning an empty array means we cannot distinguish w/ an empty directory

jsimsa replied:
sgtm

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -44,36 +34,135 @@
+      private final ConcurrentHashMap<Key, UnderFileSystem> mUnderFileSystemMap =
+          new ConcurrentHashMap<>();
+
+      Cache() {}
```
`private`?

jsimsa commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -367,17 +318,17 @@ public OutputStream create(String path, CreateOptions options) throws IOExceptio
    *
    * @param path the file name
-   * @param offset the offset in bytes
+   * @param options including offset within the file
```
it is fine to just say "method options" so that the javadoc does not need to be update / become stale when new options are added

jsimsa commented on core/common/src/main/java/alluxio/underfs/options/FileLocationOptions.java:
```diff
@@ -0,0 +1,84 @@
+/*
```
add a unit test for this class

jsimsa commented:
@madanadit left a couple more minor comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12416/
Test PASSed.

madanadit commented:
@jsimsa I added unit tests. Thanks!.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12417/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12418/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12419/<h2>Failed Tests: <span class='status-failure'>4</span></h2><h3><a name='org.alluxio:alluxio-core-common' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12419/org.alluxio$alluxio-core-common/testReport'>org.alluxio:alluxio-core-common</a>: <span class='status-failure'>4</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12419/org.alluxio$alluxio-core-common/testReport/alluxio.underfs.options/CreateOptionsTest/equals/'><strong>alluxio.underfs.options.CreateOptionsTest.equals</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12419/org.alluxio$alluxio-core-common/testReport/alluxio.underfs.options/DeleteOptionsTest/equals/'><strong>alluxio.underfs.options.DeleteOptionsTest.equals</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12419/org.alluxio$alluxio-core-common/testReport/alluxio.underfs.options/FileLocationOptionsTest/equals/'><strong>alluxio.underfs.options.FileLocationOptionsTest.equals</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12419/org.alluxio$alluxio-core-common/testReport/alluxio.underfs.options/MkdirsOptionsTest/equals/'><strong>alluxio.underfs.options.MkdirsOptionsTest.equals</strong></a></li></ul>
Test FAILed.

calvinjia commented on underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java:
```diff
@@ -526,219 +212,126 @@ private boolean copy(String src, String dst) {
+        throws IOException {
+      mChunk = chunk;
+      if (mChunk == null) {
+        throw new IOException("S3N listing result is null");
```
`GCS`?

calvinjia commented on underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java:
```diff
@@ -458,239 +148,146 @@ private boolean copy(String src, String dst) {
+      mRequest = request;
+      mResult = result;
+      if (mResult == null) {
+        throw new IOException("S3A listing result is null");
```
`OSS`?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12420/
Test PASSed.

jsimsa commented on core/common/src/test/java/alluxio/underfs/options/CreateOptionsTest.java:
```diff
@@ -86,7 +86,7 @@ public void fields() {
 
   @Test
-  public void equals() throws Exception {
+  public void equalsTest() throws Exception {
```
why is that?

madanadit replied:
equals seems to conflict with the java object equals method

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12421/
Test PASSed.

calvinjia commented:
LGTM
